### DISCUSSION TITLE
How to Ensure Library B Uses the Same Version of Library C as My App

### DISCUSSION THREAD

**Discussion by maliknajjar** [denoland/deno#27447] [Q&A]
Hello,

I have a scenario where my application imports Library B, and both my app and Library B rely on Library C. I want to ensure that Library B uses the same version of Library C as my application.

How can I enforce this in Deno?

**Comment by marvinhagemeister** [comment]
This is solved by both libraries specifying a version range instead of an exact version. Example:

- `lib-b` depends on `lib-c@^1.0.0` (note the `^` character)
- your app depends on `lib-c@^1.1.0` (again, note the `^` character)

This will install `lib-c` only once and both your app and `lib-b` will share the exact same version. This will always be the case for as long as the version ranges match. A scenario where they would not match would be where your app depends on `lib-c@^3.0.0` but `lib-b` depends on `lib-c@^1.0.0`. No version of `lib-c` can satisfy both version constraints, so you'll end up with two copies.

To see all available  version constraint characters, check out [this page in the docs](https://docs.deno.com/runtime/fundamentals/modules/#package-versions). Note that version specifiers only work with `jsr:` and `npm:` imports and not with `http:` or `https:`.

In the Node ecosystem you'll come across the term `peerDependencies` quite a bit which is kinda the same as version constraint, with the additional feature that package managers print a big warning or error when the version ranges don't match. Deno supports this, when you're defining your dependencies in `package.json` as well. That said for Deno first projects that is usually not needed and specifying correct version ranges is enough.

**Comment by maliknajjar** [comment]
hi @marvinhagemeister. thanks for responding

the issue with the current way is that because deno doesn't throw an error on mismatching versions. deno will end up downloading a different versions of the C library for the B library without the user even noticing it. he will do it in a silent way. no errors not even showing a warning. and this behavior can cause unexpected behavior and bugs. I actually had a bug because deno made library B use a different version of library C than the main app uses
I'm building a framework. so I have multiple libraries that are used together
and if I the author of the libraries didn't notice deno using another version of C library imagine the normal user
I can fix this now by matching the versions. but what if a user of the framework will upgrade the core app and forgets to upgrade the B library. a mismatch will happened an it will cause a lot of unexpected behavior. that's why I suggest to have the peerDependancy in the deno.json also. so that it will at least tell the user that there is a mismatch. and since peerDependancy is already implemented in deno for package.json it should be easily added in deno.json too. (Im not sure. but I think so)

**Reply by marvinhagemeister** [reply]
Right, that info that you're asking for peer dependencies explicitly was missing in your original post. There is an issue for native `peerDependencies` support in Deno here https://github.com/denoland/deno/issues/26199

**Reply by maliknajjar** [reply]
yes Im the creator of that issue 🥲

**Reply by maliknajjar** [reply]
but they tagged it as a question. not as a feature request

**Reply by marvinhagemeister** [reply]
Right, I changed the tags to match.
